src/protosym/core/sym.py

```diff
@@ -0,0 +1,615 @@
+"""The :class:`Sym` class.
+
+This module defines the :class:`Sym` class which should be used as the
+superclass for all user-facing symbolic classes. The :class:`Sym` class only
+has a handful of attributes and methods that are used to make it compatible
+with the rest of the machinery defined in `protosym.core`. The idea here is to
+build a nicer syntax over the lower-level classes that can be inherited for use
+by user-facing classes that derive from :class:`Sym`.
+"""
+from __future__ import annotations
+
+from typing import Any
+from typing import Callable
+from typing import Generic
+from typing import Optional
+from typing import overload
+from typing import Sequence
+from typing import Type
+from typing import TypeVar
+from weakref import WeakValueDictionary as _WeakDict
+
+from protosym.core.atom import AtomType
+from protosym.core.evaluate import Evaluator
+from protosym.core.exceptions import BadRuleError
+from protosym.core.tree import TreeAtom
+from protosym.core.tree import TreeExpr
+
+__all__ = ["Sym", "SymAtomType", "SymEvaluator"]
+
+T_sym = TypeVar("T_sym", bound="Sym")
+T_val = TypeVar("T_val")
+S_val = TypeVar("S_val")
+T_op = TypeVar("T_op")
+
+
+Wild = AtomType("Wild", str)
+
+
+class SymAtomType(Generic[T_sym, T_val]):
+    """Wrapper around AtomType to construct atoms as Sym."""
+
+    name: str
+    sym: Callable[[TreeAtom[T_val]], T_sym]
+    atom_type: AtomType[T_val]
+
+    def __init__(
+        self, name: str, sym: Callable[[TreeAtom[T_val]], T_sym], typ: Type[T_val]
+    ) -> None:
+        """New SymAtomType."""
+        self.name = name
+        self.sym = sym
+        self.atom_type = AtomType(name, typ)
+
+    def __repr__(self) -> str:
+        """The name of the SymAtomType."""
+        return self.name
+
+    def __call__(self, value: T_val) -> T_sym:
+        """Create a new Atom as a Sym."""
+        atom = self.atom_type(value)
+        return self.sym(TreeAtom[T_val](atom))
+
+    def __getitem__(self, wild: T_sym) -> SymAtomValue[T_sym, T_val]:
+        """Represent extracting the value from self."""
+        return SymAtomValue(self, wild)
+
+
+class SymAtomValue(Generic[T_sym, T_val]):
+    """Representation of extracting a value from an Atom."""
+
+    def __init__(self, atom_type: SymAtomType[T_sym, T_val], wild: T_sym) -> None:
+        self.atom_type = atom_type
+        self.args = (wild,)
+
+
+class Sym:
+    """Base class for user-facing symbolic classes.
+
+    This class should not be used directly but rather subclassed to make a
+    user-facing symbolic expression type:
```

Comment:
If this is the case could we have it subclass `abc.ABC` from the standard library to enforce this behaviour?

Reply:
I'm not sure. I seem to remember that the ABCs are slow or that they slow down isinstance testing.

Right now it's just difficult to use Sym directly because it doesn't provide the methods that you would want. The main missing method is `__call__` but I didn't want to export that to all potential subclasses so I didn't add it. It already has too many attributes and methods now.

I've never really used ABC so I'm not sure exactly what the benefits are. Does it just mean that you can't create instances?

Reply:
We could also add something to make Sym easier to use although I'm not sure what the usecase would be for using it directly. It might be useful for Sym (and all subclasses) to accept a tuple as an argument like:
```python
expr = Sym((cos, one))
```
That would at least make it a bit easier to create a `Sym` instance making it potentially more useful.

Reply:
> I've never really used ABC so I'm not sure exactly what the benefits are. Does it just mean that you can't create instances?

It lets you specify which methods a concrete class must implement for it to be able to create instances. For example, if you wanted to ensure that all subclasses of `Sym` implement their own `__init__` method then you could do something like

```python
from abc import ABC, abstractmethod

class Sym(ABC):

    @abstractmethod
    def __init__(self):
        pass
```

Then if a user tries to create an instance of `Sym` directly they'll get an error: `TypeError: Can't instantiate abstract class Sym with abstract method __init__`

Similarly, if a user tries to implement a concrete subclass of `Sym`, e.g. `Expr`, without overriding the `__init__` method : `TypeError: Can't instantiate abstract class Expr with abstract method __init__`

But this is only useful if there are specific methods/properties that you want to enforce have to be overridden. If there aren't and you just want to ensure that `Sym` is never instantiated then you'd likely need to do something like

```python
class Sym:

    def __new__(cos, *args, **kwargs):
        if cls is BaseClass:
            raise TypeError(f"'{cls.__name__}' can't be instantiated directly")
        return object.__new__(cls, *args, **kwargs)
```

But in this case, just documenting that `Sym` is intended to be subclassed, as you've done, is probably the best approach.

I also don't know whether there's any performance penalty associated with having the base class inherit from `abc.ABC`.